[docs] Add writing style guide (#4271)
drud/ddev: docs/content/developers/writing-style-guide.md

## Voice and Tone

In the spirit of the [Code of Conduct](../users/code-of-conduct.md), we want to be clear and encouraging for everyone that bothers to read DDEV’s documentation, rewarding the time and attention they choose to give to it.

### Beginner-Friendly, Expert-Compatible

Write so a DDEV beginner can follow your guidance and a DDEV veteran could use the same content as a reference.

| Write This 👍 | Not This ❌
| -- | --
| You can run `ddev restart` to apply changes you’ve made to your global `~/.ddev/global_config.yaml` or per-project `.ddev/config.yaml`. | Just `ddev restart` to apply YAML config changes.

### Be Efficient and Direct

Brevity makes for good documentation!

Few read the docs from start to finish like a book, and explanation can be a barrier to learning. Get to the point, avoiding parentheticals and mid-sentence notes that interrupt the main flow.

Omit extraneous explanation or decorative language that doesn’t help the reader. Instructions don’t need to include “please”. Provide some context for anchor links without overloading them to hinder flow.

| Write This 👍 | Not This ❌
| -- | --

> | Run `ddev start` and launch the site in a browser. | Please run `ddev start`, then launch the site in a browser.

Reviewer: ```suggestion
| Run `ddev start` and launch the site in a browser. | Please run `ddev start`, then use `ddev launch` to view the site in a browser.
```

Author: This may be an awful example to start with—added `ddev launch` to the _ideal_ example as well, since your suggested improvement only made the “Not This” example more useful.

```suggestion
| Run `ddev start`, then run `ddev launch` to open the site in a browser. | Please run `ddev start`, then use `ddev launch` to view the site in a browser.
```

Author: This example is really just illustrating how not to start a sentence with “Please”, which is probably insulting to anyone with a pulse. Might be better to get rid of it or come up with another one.

Reviewer: Your choice here. It's hard to review these tables in a PR, so gets confusing.

Author: I’ll come back to this and try for a new example. In the meantime, I added a screenshot to the PR description since the illustrations (via table and `!!!tip` examples) are a big part of it.

Reviewer 2: @rfay 
![image](https://user-images.githubusercontent.com/8931305/194672710-7b34524d-2b13-49d7-937e-31872087dbf7.png)

Clicking on this button will give you a more Visual view of the PR.

| Learn more on the [Extending](../users/extend/customization-extendibility.md) page. | (You can also learn more about this and related topics in [Providing Custom Environment Variables to a Container](../users/extend/customization-extendibility.md).)

### Avoid “Just” and “Easy”

> Try not to use language that may talk down to the reader. You may intend for “it’s easy” to be reassuring, but it’s a subjective judgment that can convince someone struggling that they’re doing it wrong. Things could instead be “straightforward” if they’re without nuance, “simple” if they don’t involve complex actions or concepts, or “quick” if they involve one or two steps that’d be fast even on someone’s worst day with the slowest-imagineable machine.

Reviewer: Oddly, the docs check didn't get run, which should have caught this. We'll see if it runs when you commit this suggestion.

> Try not to use language that may talk down to the reader. You may intend for “it’s easy” to be reassuring, but it’s a subjective judgment that can convince someone struggling that they’re doing it wrong. Things could instead be “straightforward” if they’re without nuance, “simple” if they don’t involve complex actions or concepts, or “quick” if they involve one or two steps that’d be fast even on someone’s worst day with the slowest-imagineable machine.

Author: I got in the habit of using `[skip ci]` so that’s probably my fault. Thanks @mayankguptadotcom!

> Try not to use language that may talk down to the reader. You may intend for “it’s easy” to be reassuring, but it’s a subjective judgment that can convince someone struggling that they’re doing it wrong. Things could instead be “straightforward” if they’re without nuance, “simple” if they don’t involve complex actions or concepts, or “quick” if they involve one or two steps that’d be fast even on someone’s worst day with the slowest-imagineable machine.

Reviewer: For PRs that are just docs, there's no need to use [skip ci] because most of the tests are skipped when the system discovers it's just docs. I imagine you already know that.

> Try not to use language that may talk down to the reader. You may intend for “it’s easy” to be reassuring, but it’s a subjective judgment that can convince someone struggling that they’re doing it wrong. Things could instead be “straightforward” if they’re without nuance, “simple” if they don’t involve complex actions or concepts, or “quick” if they involve one or two steps that’d be fast even on someone’s worst day with the slowest-imagineable machine.

Author: I did not, so thank you!

| We can use “curly quotes” now that we’re post-typewriter. | We can use "curly quotes" now that we're post-typewriter.

### “Run” Commands

We “run” commands. We don’t “do” them, and the command itself is not a verb. Whenever possible, reinforce that a given thing in backticks is intended as a console command by using the word “run” before it.

| Write This 👍 | Not This ❌ |
| -- | -- |
| Run `ddev config` to set up your project. | Do `ddev config` to set up your project.<br>You can `ddev config` to set up your project. |
| If you get stuck, run `ddev restart`. | If you get stuck, just `ddev restart`. |

### Use Active Third Person

Avoid impersonal language featuring unknown individuals or shadowy organizations.  
“It is recommended,” for example, could be a warmer “we recommend” or “Laravel users recommend”.

Write on behalf of the community and not yourself—use “we” and not “I”.

| Write This 👍 | Not This ❌ |
| -- | -- |
| We recommend Colima for the best performance. | It is recommended for performance that you use Colima. |
| Some prefer Redis for runtime caches. | I like using Redis for runtime caches. |

### Write Once and Link

> Try to keep from repeating yourself in the documentation. Instead, write carefully and link to that well-crafted specimen, whether it’s across the page or off to another section. This has two benefits:

Reviewer: Absolutely correct, but there are times that links with necessary info distract from the flow, so it's better to just tell them the missing info.

Author: The realist in me probably agrees, but I have lists of redundancies to go through and address—both linking to canonical explanation and adding the _ability_ to link it in the first place. (Like individual commands that live in a table, for example.)

1. Easier maintenance with less chance of redundant information becoming stale.
2. Subtle reinforcement of documentation structure that helps the reader learn where to find answers, rather than answering the same thing in different places.

### Mind Your Context

It’s easy to get lost in documentation; don’t assume the reader is always following your words. Take care to bring the reader with you, especially if there are steps that involve different applications or distinct areas of concern.

| Write This 👍 | Not This ❌ |
| -- | -- |
| 1. In Docker Desktop, go to *Resources* → *Advanced* and set “Memory” to 6GB.<br>2. From your terminal, run `ddev restart`.<br>3. In your text editor, open `.ddev/config.yaml` and set `php_version: "8.1"`. | 1. Go to *Resources* → *Advanced* and set “Memory” to 6GB.<br>2. Run `ddev restart`.<br>3. Set `php_version: "8.1"`. |

> | Once you’ve [installed a Docker provider](docker-installation.md), you’re ready to install DDEV! | Docker or an alternative is required before anything will work with DDEV. This is pretty easy on most environments; see the [Docker Installation](../users/install/docker-installation.md) page to help sort out the details.

Reviewer: ```suggestion
| Once you’ve [installed a Docker provider](../users/install/docker-installation.md), you’re ready to install DDEV! | Docker or an alternative is required before anything will work with DDEV. This is pretty easy on most environments; see the [Docker Installation](../users/install/docker-installation.md) page to help sort out the details.
```

| PostgreSQL | Postgres
| Terminus | terminus
| Windows | windows
| Xdebug | XDebug, xDebug

### Quote Copied Text

If you’re quoting a human being or a message lifted verbatim from some other source (outside a fenced code block), make sure it ends up in a `<blockquote>` element:

```
> Error: your quote style should not always be in a fenced block.
```

> Error: your quote style should not always be in a fenced block.

### Other Recommendations

One-off tips that don’t fit nicely into any of the sections above:

- Pluralize “CMS” as “CMSes”, not “CMSs”.
- Use all-caps references for file *types* like JSON, YAML and CSS.
- Wrap file *extensions* in backticks like `.json`, `.yaml`, and `.css`.
- Wrap references to files, directories, images and commands in backticks.
- Use Title Case for headings wherever it makes sense.
- Link to related services and topics where convenient—usually first use on a given page.
- Use `<kbd>` elements for representing literal keystrokes.
- Use sequential numbers for numbered lists in the source Markdown, regardless of how they’re eventually rendered.
- Try to maintain parallel format for list items.

| Write This 👍 | Not This ❌ |
| -- | -- |

> | web server | webserver

Reviewer: I like webserver better :)

Author: You’re clearly not alone! Same issue in [this ancient battleground](https://code.djangoproject.com/ticket/20330).

Reviewer: I can live with styles I don't naturally do.
